internal/client.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	return nil
+}
+
+func (c *Client) HeadBucket(ctx context.Context, bucketName string) (bool, error) {
+	_, err := c.s3Client.HeadBucket(ctx, &s3.HeadBucketInput{
+		Bucket: aws.String(bucketName),
+	})
+
+	exists := true
+	if err != nil {
+		var notFoundErr *s3types.NotFound
+		if ok := errors.As(err, &notFoundErr); ok {
+			exists = false
+			return exists, nil
+		}
+	}
+
+	return exists, err
+}
+
+func (c *Client) DeleteBucket(ctx context.Context, bucketName string) error {
+	_, err := c.s3Client.DeleteBucket(ctx, &s3.DeleteBucketInput{
+		Bucket: aws.String(bucketName),
+	})
+
+	return err
+}
+
+func (c *Client) GetBucketMetadata(ctx context.Context, bucketName string) (*types.BucketMetadata, error) {
+	params := map[string]string{
+		"metadata": "",
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.bucketURL(bucketName, params), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %v", err)
+	}
+
+	resp, err := c.doRequestWithRetry(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send request: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("request failed with code: %d", resp.StatusCode)
+	}
+
+	// Parse the response body into a BucketMetadata struct
+	var metadata types.BucketMetadata
+	err = json.NewDecoder(resp.Body).Decode(&metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read bucket metadata: %v", err)
+	}
+
+	return &metadata, nil
+}
+
+func (c *Client) doRequestWithRetry(req *http.Request) (*http.Response, error) {
+	maxRetries := 5
+	backoffDelay := 3 * time.Second
+	maxBackoffDelay := 60 * time.Second
+
+	var resp *http.Response
+	var err error
+
+	for i := 0; i < maxRetries; i++ {
+		// Clone the request to avoid issues with mutated request objects
+		clonedReq, err := cloneRequest(req)
+		if err != nil {
+			return nil, fmt.Errorf("failed to clone request: %v", err)
+		}
+
+		resp, err = c.doSignedRequest(clonedReq)
+		if err != nil {
+			return nil, fmt.Errorf("failed to send request: %v", err)
+		}
+
+		// Check if the response status code indicates a server-side error (5xx)
+		if resp.StatusCode >= 500 {
+			resp.Body.Close()
+
+			// Exponential backoff before retrying
+			time.Sleep(backoffDelay)
+			backoffDelay *= 2 // Double the delay for each retry
+			if backoffDelay > maxBackoffDelay {
+				backoffDelay = maxBackoffDelay
+			}
+
+			continue
+		}
+
+		// Break out of the loop if the request was successful
+		break
+	}
+
+	return resp, err
+}
+
+func (c *Client) doSignedRequest(req *http.Request) (*http.Response, error) {
+	// Sign the request
+	err := c.signRequest(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign request: %v", err)
+	}
+
+	// Send the signed request using the wrapped http.Client
+	return c.httpClient.Do(req)
+}
+
+func (c *Client) bucketURL(bucketName string, queryParams map[string]string) string {
+	baseURL := fmt.Sprintf("%s/%s", c.endpoint, bucketName)
```

Comment:
this will work for all cases except the custom domain configured buckets.

Reply:
This should work with custom domain buckets as well, as custom domains are just an additional way to access the bucket. Similar to how it works with S3 client for such buckets.

Reply:
bucket name is embedded in domain /<bucket-name> name will not work

Reply:
Both will work: images.example.com, and fly.storage.tigris.dev/images.example.com

Reply:
Yes, it will work. Even with custom domain the path prefix works.